### PR TITLE
#110 sipservlet would have the feature to automatically quote parameter value if parameter name is in defined quoted parameters list

### PR DESCRIPTION
**Pull request by xhoaluu**
Hi, please help me to review and give me comment to fix if it's not good about the code.

BR
Hoan HL

### CODE REVIEW

**Comment by xhoaluu** [comment]
Hi @jaimecasero,

I have finished implementation for #110. Could you help me a bit on reviewing the code and give me comment :)

File: sip-servlets-test-suite/applications/shootist-sip-servlet/src/main/java/org/mobicents/servlet/sip/testsuite/ShootistSipServlet.java [hunk 1]
```diff
@@ -439,6 +439,12 @@ public void servletInitialized(SipServletContextEvent ce) {
 				addr.setParameter("headerparam1", "headervalue1");
 				addr.setParameter("param5", "ffff");
 				addr.getURI().setParameter("uriparam", "urivalue");
+				if (ce.getServletContext().getInitParameter("testContactHeader") != null){
+					addr.setParameter("param0", "value0");
```
Comment (jaimecasero) [inline sip-servlets-test-suite/applications/shootist-sip-servlet/src/main/java/org/mobicents/servlet/sip/testsuite/ShootistSipServlet.java]: this is what i meant on resuign existing test, good job!!

File: sip-servlets-impl/src/main/java/org/mobicents/servlet/sip/address/ParameterableImpl.java [hunk 2]
```diff
@@ -154,8 +158,29 @@ public void setParameter(String name, String value) {
...
 		this.parameters.put(name.toLowerCase(), value);
 		if(header != null) {
+			boolean isQuotableParameter = false;
+			SipContext context = SipApplicationSessionCreationThreadLocal.lookupContext();
+			if (context != null){
+				Object object = context.getServletContext().getAttribute("org.restcomm.servlets.sip.QUOTABLE_PARAMETER");
+				if (object != null){
+					String quotableParameters = object.toString();
+					if (!quotableParameters.isEmpty()){
+						String[] parameters = quotableParameters.split(",");
```
Comment (jaimecasero) [inline sip-servlets-impl/src/main/java/org/mobicents/servlet/sip/address/ParameterableImpl.java]: could we isolate this logic in a private method, something like isQutoableParam()?

File: sip-servlets-test-suite/applications/shootist-sip-servlet/src/main/java/org/mobicents/servlet/sip/testsuite/ShootistSipServlet.java (hunk 1 above)
Comment (xhoaluu) [inline sip-servlets-test-suite/applications/shootist-sip-servlet/src/main/java/org/mobicents/servlet/sip/testsuite/ShootistSipServlet.java]: Hi @jaimecasero,
I have updated the code based on your comments. Please take a look again.

I also have pulled a request for jain-sip to support #110, please help me to review it as well.
Link: https://github.com/RestComm/jain-sip/pull/106

File: sip-servlets-impl/src/main/java/org/mobicents/servlet/sip/address/ParameterableImpl.java [hunk 3]
```diff
@@ -155,13 +159,44 @@ public void setParameter(String name, String value) {
...
+	 * Check if the param name is in the list that its value need to be quoted
+	 * 
+	 * @param name - a string specifying the parameter name
+	 * 
+	 */
+	private boolean isQuotableParam(String name){
+		boolean isQuotableParameter = false;
+		SipContext context = SipApplicationSessionCreationThreadLocal.lookupContext();
+		if (context != null){
+			Object object = context.getServletContext().getAttribute("org.restcomm.servlets.sip.QUOTABLE_PARAMETER");
```
Comment (jaimecasero) [inline sip-servlets-impl/src/main/java/org/mobicents/servlet/sip/address/ParameterableImpl.java]: could we build a default value for this attribute, including known attributes from SIP RFCs?

**Comment by xhoaluu** [comment]
I have tried to list out the list of generic parameters that their values need to be quoted: 
"vendor, model, version, nextnonce, nonce, code, Identity, oc-algo, cid, text". Please help me to check those params.

Please take a look at SipStandardContext.java as well.

**Comment by jaimecasero** [comment]
@xhoaluu  on my side I found these params to be quoted on RFC2161 basic rules:
cnonce,domain,nextnonce,nonce,opaque,qop,realm,response,rspauth,uri,username

Some of the ones you mentioned I wasnt able to locate in RFC3261, where are they coming from?

**Comment by xhoaluu** [comment]
for vendor, model, version:
[http://www.tech-invite.com/y60/tinv-ietf-rfc-6080-3.html#e-6-2-2]("vendor", "model", and "version" Parameters)
[http://www.tech-invite.com/fo-abnf/tinv-fo-abnf-sip-h-geolocerr.html#location-error-code-text](code)
Identity is a header, not a header param, remove it.
[http://www.tech-invite.com/y70/tinv-ietf-rfc-7339.html#e-9]( oc-algo)
[http://www.tech-invite.com/fo-abnf/tinv-fo-abnf-sip-h-triggcons.html#hd]("target-uri"   EQUAL   LDQUOT   *( qdtext  /  quoted-pair )   RDQUOT)
[http://www.tech-invite.com/fo-abnf/tinv-fo-abnf-sip-h-secclient.html#hd]("d-ver"   EQUAL   LDQUOT   32LHEX   RDQUOT)

**Comment by jaimecasero** [comment]
ok, can you merge both list into the code?. I think after that change we can finally merge the code!!!

**Comment by xhoaluu** [comment]
hi @jaimecasero ,
I updated the known parameter list. All the things are ready now.

**Comment by jaimecasero** [comment]
congrats!!! @xhoaluu   PR merged into master. we will trigger CI job now, and check testsuite

**Comment by xhoaluu** [comment]
thanks so much @jaimecasero, It took me quite long time to finish but at the end, the feeling is so happy :)

**Pull request merged**